---
title: "Figure 1, Isolates Paper"
author: "Dylan Baker"
output:
  html_document:
    code_folding: hide
    highlight: haddock
    keep_md: yes
    number_sections: yes
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  pdf_document:
    toc: yes
---
```{r packages}
require("knitr")

list.of.packages <- c(
  "tidyverse",
  "ggpubr",
  "patchwork",
  "tidytext"
)
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])] 
if(length(new.packages)) install.packages(new.packages)

library(tidyverse)
library(ggpubr)
library(patchwork)
library(tidytext)


```

``` {r Import Data}
Phycosphere_tax <- read_tsv(file="../mothur_outputs/all_seqs/final.merge.asv.ASV.cons.taxonomy") |>
		rename_all(tolower) |>  mutate(taxonomy=str_replace_all(string=taxonomy, pattern="\\(\\d*\\)", replacement="")) |>
		mutate(taxonomy=str_replace_all(string=taxonomy , pattern=";$", replacement="")) |>
    rename("asv" = otu) |>
		separate(taxonomy, into=c("domain", "phylum", "class", "order", "family", "genus"), sep=";")
	
Phycosphere_asv <- read_tsv("../mothur_outputs/all_seqs/final.merge.asv.shared", col_types=cols(Group=col_character())) |>
	select(-label, -numASVs) |>
 	pivot_longer(cols=-Group, names_to="asv", values_to="count")|>
  filter(count > 0)

sampledata <- read_csv("../mothur_outputs/all_seqs/GRBC_metadata.csv")

#Create two separate dataframes with phycosphere and isolate data.
phycosphere_data <- anti_join(Phycosphere_asv, sampledata, by = "Group")

GRBC_data <- inner_join(Phycosphere_asv, sampledata, by = "Group") |>
  mutate(color = case_when(host_species == "Chlorella" ~ "#7FC97F",
          host_species == "Coelastrum"~ "#BEAED4",
          host_species == "Scenedesmus"~ "#E0115F",
          host_species == "Monoraphidium"~"#FDC086",
          host_species == "Selenastrum"~"#386CB0", 
          TRUE ~ "black"))

GRBC_counts <- GRBC_data |> 
  group_by(Group) |>
  summarize(n = sum(count)) |>
  ungroup() |>
  rename("total_count_group" = n)
                              
GRBC_data <- inner_join(GRBC_data, GRBC_counts, by = "Group") |>
  mutate(Isolate_Number = case_when(seq_type == "Sanger" ~ str_remove(Isolate_Number,"_"), 
                                 T ~ Isolate_Number),
           rel_abund = count/total_count_group)

#We need the number from this output to standardize counts below.
Phycosphere_counts <- phycosphere_data |>
  #make the data table recognize samples
  group_by(Group) |>
  summarize(n = sum(count)) |>
  ungroup() |>
  rename("total_count_group" = n)

Phycosphere_asv_count <- inner_join(Phycosphere_asv, Phycosphere_counts, by = "Group") |>
  mutate(rel_abund = count/total_count_group) |>
  #Add Host Species Information
    mutate(host_species = case_when(str_detect(Group, "Chlorella") ~ "Chlorella",
                        str_detect(Group, "Coelastrum") ~ "Coelastrum",
                        str_detect(Group, "Scenedesmus") ~ "Scenedesmus",
                        str_detect(Group, "Monoraphidium") ~ "Monoraphidium",
                        str_detect(Group, "Selenastrum") ~ "Selenastrum",
                        TRUE ~"Natural Community"))

Phycosphere_asv_D3 <- Phycosphere_asv_count |> 
  filter((str_detect(Group,"D3$")))


Phycosphere_asv_D31 <- Phycosphere_asv_count |> 
  filter(str_detect(Group,"D31"))

```


```{r Phycosphere OTU D3/D31 Filtering 0.22 um}
#Separate D3 and D31 data
D3_data <- inner_join(Phycosphere_asv_D3, Phycosphere_tax) |> #Groups by asv automatically
  mutate(time = "D3",
         genus=str_replace_all(string=genus,
               pattern="Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium", replacement="A-N-P-R"))

D31_data <- inner_join(Phycosphere_asv_D31, Phycosphere_tax) |> #Groups by asv automatically
  mutate(time = "D31",
         genus=str_replace_all(string=genus,
               pattern="Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium", replacement="A-N-P-R")) 

#Recombine as one dataframe for contamination checks
agg_genus_data <- rbind(D3_data, D31_data)
# unique genera
length(unique(agg_genus_data$genus)) 

```

```{r GRBC Comparison}

#Filter genus hits from GRBC silva and freshwater database hits
Phycosphere_GRBC_D3 <- inner_join(GRBC_data, D3_data, by = c("asv", "host_species")) |>
  filter(capture_day == "D3") |>
  select(asv, host_species, color) |>
  distinct() 

Phycosphere_GRBC_D31 <- inner_join(GRBC_data, D31_data, by = c("asv", "host_species")) |>
  filter(capture_day == "D31")  |>
  select(asv, host_species, color) |>
  distinct()


#Isolates that are probably contaminants, not represented in phycosphere data
Phycosphere_GRBC_cont <- anti_join(GRBC_data, agg_genus_data, by = "asv") |>
  mutate(contaminated = TRUE) |> 
  filter(rel_abund > 0.02) |> #filter out asvs that are likely sequencing error or true contamination.
  inner_join(Phycosphere_tax, by = "asv")
cont_summary <- Phycosphere_GRBC_cont |> group_by(genus) |>
  summarize(distinct_isolate = n_distinct(Group),
            distinct_asv = n_distinct(asv))
write.csv(Phycosphere_GRBC_cont, file = "Isolate_contamination_report.csv")
write.csv(cont_summary, file = "Isolate_contamination_summary.csv")

agg_genus_data_D3_color <- D3_data |> 
  #Filter out reads below a threshold (1/500 worked best) that are not part of our isolate collection
  filter(!(rel_abund < 0.01) | asv %in% Phycosphere_GRBC_D3$asv | asv %in% Phycosphere_GRBC_D31$asv) |>
  left_join(Phycosphere_GRBC_D3, by = c("asv", "host_species")) |>
  group_by(asv) |>
  mutate(color = if_else(condition = is.na(color), true = "black", false = color),
         subset = ifelse(rel_abund > 0.01, "Abundant","Rare (<1% of Total Reads Within a Host Phycosphere)"),
         significant = ifelse(color == "black", "", "*"),
         asv = str_replace(string = asv, pattern = "ASV0+", replacement = ''),
         time = "D3",
         host_species = factor(host_species, 
                               levels = c("Chlorella", "Coelastrum", "Scenedesmus", "Monoraphidium", 
                                          "Selenastrum", "Natural Community"))) |>
  ungroup()

agg_genus_data_D31_color <- D31_data |> 
  #Filter out reads below the detection limit that are not part of our isolate collection
  filter(!(rel_abund < 0.01) | asv %in% Phycosphere_GRBC_D3$asv | asv %in% Phycosphere_GRBC_D31$asv) |>
  left_join(Phycosphere_GRBC_D31, by = c("asv", "host_species")) |>
  mutate(color = if_else(condition = is.na(color), true = "black", false = color),
         significant = ifelse(color == "black", "", "*"),
         asv = str_replace(string = asv, pattern = "ASV0+", replacement = ''),
         time = "D31",
         host_species = factor(host_species, 
                               levels = c("Chlorella", "Coelastrum", "Scenedesmus", "Monoraphidium", 
                                          "Selenastrum", "Natural Community"))) |>
  ungroup()



D3_plots_combined <- agg_genus_data_D3_color |>
  ggplot(aes(
    x = reorder_within(asv,-rel_abund, fun = sum, within = subset),
    y = rel_abund,
    fill = host_species
  )) +
  geom_col() +
  scale_x_reordered() +
  facet_wrap( ~ subset, scales = "free") +
  theme_pubclean() +
  scale_fill_manual(
    labels = c(
      "Chlorella" = expression(italic("C. sorokiniana")),
      "Coelastrum" = expression(italic("C. microporum")),
      "Scenedesmus" = expression(italic("S. acuminatus")),
      "Monoraphidium" = expression(italic("M. minutum")),
      "Selenastrum" = expression(italic("S. capricornutum"))
    ),
    values = c(
      "Chlorella"  = "#7FC97F",
      "Coelastrum" = "#BEAED4",
      "Scenedesmus" = "#E0115F",
      "Monoraphidium" = "#FDC086",
      "Selenastrum" = "#386CB0"
    ),
    "Host Algae"
  )  +
  geom_text(
    aes(label = significant),
    position = position_stack(vjust = 0.5),
    vjust = .8,
    size = 20 / .pt
  ) +
  theme(axis.text.x = element_text(
    angle = 90,
    vjust = 0.5,
    hjust = 1
  )) +
  labs(y = "Relative Abundance (ASV)",
       x = "",
       fill = '',
       title = "Rank Abundance Curves of Green Alga Phycospheres, 3 Days Post Introduction to Pond Water")


D31_plots_combined <- agg_genus_data_D31_color |>
  mutate(
    subset = ifelse(
      rel_abund > 0.01,
      "Abundant",
      "Rare (<1% of Total Reads Within a Host Phycosphere)"
    )
  ) |>
  ggplot(aes(
    x = reorder_within(asv,-rel_abund, fun = sum, within = subset),
    y = rel_abund,
    fill = host_species
  )) +
  geom_col() +
  scale_x_reordered() +
  facet_wrap( ~ subset, scales = "free") +
  theme_pubclean() +
  scale_fill_manual(
    labels = c(
      "Chlorella" = expression(italic("C. sorokiniana")),
      "Coelastrum" = expression(italic("C. microporum")),
      "Scenedesmus" = expression(italic("S. acuminatus")),
      "Monoraphidium" = expression(italic("M. minutum")),
      "Selenastrum" = expression(italic("S. capricornutum"))
    ),
    values = c(
      "Chlorella"  = "#7FC97F",
      "Coelastrum" = "#BEAED4",
      "Scenedesmus" = "#E0115F",
      "Monoraphidium" = "#FDC086",
      "Selenastrum" = "#386CB0"
    ),
    "Host Algae"
  )  +
  geom_text(
    aes(label = significant),
    position = position_stack(vjust = 0.5),
    vjust = .8,
    size = 20 / .pt
  ) +
  theme(axis.text.x = element_text(
    angle = 90,
    vjust = 0.5,
    hjust = 1
  )) +
  labs(y = "Cumulative Relative Abundance (ASV)",
       x = "Amplicon Sequence Variant",
       fill = '',
       title = "Rank Abundance Curves of Green Alga Phycospheres, 31 Days Post Introduction to Pond Water") 

final_plot <- (D3_plots_combined / D31_plots_combined) + plot_annotation(
    tag_levels = "A"
) + plot_layout(
  guides = 'collect'  
) & theme(legend.position = "bottom",
          legend.justification = "center",
          )

png(filename = "abundance_plots D3 vs 31.png",
    res = 450,
    type = "cairo",
    units = "in",
    width = 14,
    height = 10)
final_plot
dev.off()

```

```{r Additional Analyses}
#How many ASVs out of the total phycosphere were represented in our collection?
D3_hits <- GRBC_data |>
  #filter(capture_day == "D3") |>
  select(asv, host_species) |>
  distinct(asv, host_species) |>
  inner_join(D3_data, by = c("asv", "host_species")) |>
  summarise(rep_asvs = sum(rel_abund)/(n_distinct(host_species)))


D31_hits <- GRBC_data |>
  #filter(capture_day == "D31") |>
  select(asv, host_species) |>
  distinct(asv, host_species) |>
  inner_join(D31_data, by = c("asv", "host_species")) |>
  summarise(rep_asvs = sum(rel_abund)/n_distinct(host_species)) 

#~18.4% so about a 5th of the diversity represented we have captured with our collection
all_hits <- GRBC_data |>
  select(asv, host_species) |>
  distinct(asv, host_species) |>
  inner_join(agg_genus_data, by = c("asv", "host_species")) |>
  summarise(rep_asvs = sum(rel_abund)/n_distinct(Group)) 

#Look at shared ASVs among multiple host phycospheres
all_hits <- GRBC_data |>
  select(asv, host_species) |>
  distinct(asv, host_species) |>
  inner_join(agg_genus_data, by = c("asv", "host_species")) |>
  group_by(asv) |>
  summarise(host_presence = n_distinct(host_species))
write.csv(all_hits, "asv_host_prevalence.csv")

#Look at % of asvs that are present in more than three host phycospheres at D3 that are >1% abundant
D3_shared_asvs <- D3_data |>
  filter(rel_abund > 0.01) |>
  group_by(asv) |>
  summarise(shared_hosts = n_distinct(host_species)) |>
  summarise(percent_asv = sum(shared_hosts >= 3)/n())

D31_shared_asvs <- D31_data |>
  filter(rel_abund > 0.01) |>
  group_by(asv) |>
  summarise(shared_hosts = n_distinct(host_species)) |>
  summarise(percent_asv = sum(shared_hosts >= 3)/n())

pure_isolates <- GRBC_data |>
  left_join(Phycosphere_tax, by = "asv") |>
  filter(rel_abund >= .1) |>
  group_by(Group) |>
  mutate(numasvs = n(),
         mixed = ifelse(numasvs > 1, T, F)) |>
  filter(mixed == F) |>
  ungroup() |>
  distinct(Isolate_Number, .keep_all = T) |>
  #Duplicate isolates that need to be filtered by hand.
  filter(!Group %in% c('11_DF','S15','S16','S20','S25','34_DF','S3','6_DF','7_DF','8_DF','S9W'))

frequency <- pure_isolates |>
  group_by(capture_day) |>
  summarise(n = n())

mixed_isolates <- GRBC_data |>
  left_join(Phycosphere_tax, by = "asv") |>
  filter(rel_abund >= .1) |>
  group_by(Group) |>
  mutate(numasvs = n(),
         mixed = ifelse(numasvs > 1, T, F)) |>
  filter(mixed == T) 

frequency_mixed <- mixed_isolates |>
  distinct(Group, .keep_all = T) |>
  group_by(capture_day) |>
  summarise(n = n())

duplicates <- pure_isolates |>
  group_by(asv,capture_day) |>
  summarise(distinct_hosts = n_distinct(host_species),
            distinct_isolates = n_distinct(Group)) |>
  ungroup() |>
  arrange(distinct_hosts, desc = T) |>
  pivot_wider(names_from = c(capture_day, distinct_hosts), values_from = distinct_isolates) |>
  summarise(across(D31_1:D3_5, ~sum(.x, na.rm = T))) |>
  pivot_longer(everything()) |>
  separate(name, into = c("capture_day","distinct_hosts"), sep = "_") |>
  left_join(frequency) |>
  mutate(fraction = value/n) |>
  group_by(distinct_hosts) |>
  summarise(total_fraction = mean(fraction))
  
pure_isolate_tax <- pure_isolates |>
  select(Group, Isolate_Number, asv, host_species, capture_day, domain:genus)
write.csv(pure_isolate_tax, "GRBC_pure_isolate_taxonomy.csv")
```